### PR TITLE
feat: push down hexInt/hexStr function to tiflash

### PR DESCRIPTION
**Pull request by YangKeao**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/35901

Related PR: https://github.com/pingcap/tiflash/pull/5280

Problem Summary:

Implement `hexInt`/`hexStr` function push down.

### What is changed and how it works?

TiDB Side: add `ast.Hex` to the `scalarExprSupportedByFlash` function to push down.

TiFlash Side:

- `hexStr` is manually implemented through convert every bytes to two characters.
- `hexInt` is implemented by calling `sprintf(_, "%X", _)` directly.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Support to pushdown hexInt/hexStr to TiFlash
```

### CODE REVIEW

**Comment by SeaRise** [comment]
Please add some UT. You can see this PR https://github.com/pingcap/tidb/pull/35739 as a reference.

**Comment by SeaRise** [comment]
@YangKeao 
friendly ping~ any updated?

**Comment by YangKeao** [comment]
> @YangKeao friendly ping~ any updated?

Thanks for your remind, I nearly forgot this PR :cry: . Tests have been added. PTAL :smile_cat:

File: planner/core/integration_test.go [hunk 1]
```diff
@@ -6991,3 +6991,44 @@ func TestIssue36609(t *testing.T) {
...
+	}
+	tk.MustQuery("explain select hex(a) from t;").CheckAt([]int{0, 2, 4}, rows)
+
+	rows = [][]interface{}{
+		{"TableReader_9", "root", "data:ExchangeSender_8"},
+		{"└─ExchangeSender_8", "mpp[tiflash]", "ExchangeType: PassThrough"},
+		{"  └─Projection_4", "mpp[tiflash]", "hex(test.t.b)->Column#4"},
+		{"    └─TableFullScan_7", "mpp[tiflash]", "keep order:false, stats:pseudo"},
+	}
+	tk.MustQuery("explain select hex(b) from t;").CheckAt([]int{0, 2, 4}, rows)
```
Comment (SeaRise) [inline planner/core/integration_test.go:7033]: seems to be a duplicate?

File: planner/core/integration_test.go [hunk 2]
```diff
@@ -6991,3 +6991,44 @@ func TestIssue36609(t *testing.T) {
...
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(a int, b varchar(10));")
+	tk.MustExec("insert into t values(1, 'tiflash');")
+	tk.MustExec("set @@tidb_allow_mpp=1; set @@tidb_enforce_mpp=1;")
+	tk.MustExec("set @@tidb_isolation_read_engines = 'tiflash'")
+
+	// Create virtual tiflash replica info.
+	dom := domain.GetDomain(tk.Session())
```
Comment (SeaRise) [inline planner/core/integration_test.go:7006]: You can use `store, dom, clean := testkit.CreateMockStoreAndDomain(t)` to create `dom`.

File: planner/core/integration_test.go [hunk 3]
```diff
@@ -6991,3 +6991,44 @@ func TestIssue36609(t *testing.T) {
 	tk.MustQuery("select * from t3 straight_join t4 on t3.a = t4.b straight_join t2 on t3.d = t2.c straight_join t1 on t1.a = t2.b straight_join t5 on t4.c = t5.d where t2.b < 100 and t4.a = 10;")
 	tk.MustQuery("select * from information_schema.statements_summary;")
 }
+
+func TestHexIntOrStrPushDownToTiFlash(t *testing.T) {
```
Comment (SeaRise) [inline planner/core/integration_test.go:6995]: And add test in `expression/expr_to_pb_test.go::TestExprPushDownToFlash`?

File: planner/core/integration_test.go (hunk 1 above)
Comment (SeaRise) [reply planner/core/integration_test.go:7033]: sorry, my mistake...

File: planner/core/integration_test.go (hunk 2 above)
Comment (YangKeao) [reply planner/core/integration_test.go]: Good! I will also use `dom.InfoSchema().TableByName` to get the table.

File: planner/core/integration_test.go (hunk 3 above)
Comment (YangKeao) [reply planner/core/integration_test.go:6995]: Added

**Review by SeaRise — APPROVED** [review]
LGTM

**Review by gengliqi — APPROVED** [review]
LGTM

**Review by guo-shaoge — APPROVED** [review]

**Pull request merged**